dynamic_instrumentation/src/instrument.rs

```diff
@@ -343,43 +343,110 @@ impl<'tcx> Visitor<'tcx> for CollectInstrumentationPoints<'_, 'tcx> {
+                break;
+            }
+            let _ = proj_iter.next();
+        }
+
+        while let Some((deref1_base, PlaceElem::Deref)) = proj_iter.next() {
+            // Find the next Deref or end of projections
+            let (deref2_base, have_deref2) = loop {
+                match proj_iter.peek() {
+                    Some((base, PlaceElem::Deref)) => break (base.clone(), true),
+                    // Reached the end, we can use the full place
+                    None => break (place.as_ref(), false),
+                    _ => {
+                        let _ = proj_iter.next();
+                    }
                 }
+            };
+
+            // We have some other elements between the two projections
+            if deref2_base.projection.len() - deref1_base.projection.len() > 1 {
+                self.loc(location, location, project_fn)
+                    .arg_var(deref1_base)
+                    .arg_addr_of(deref2_base.clone())
+                    .add_to(self);
+            }
+
+            use PlaceContext::*;
+            if let Some(ptr_fn) = match context {
+                // We are just loading the address for the next dereference
+                _ if have_deref2 => Some(load_fn),
+
+                NonMutatingUse(NonMutatingUseContext::Inspect) |
+                NonMutatingUse(NonMutatingUseContext::Copy) |
+                NonMutatingUse(NonMutatingUseContext::Move) => Some(load_fn),
+
+                MutatingUse(MutatingUseContext::Store) |
+                MutatingUse(MutatingUseContext::Call) |
+                MutatingUse(MutatingUseContext::AsmOutput) => Some(store_fn),
+
+                NonMutatingUse(NonMutatingUseContext::ShallowBorrow) |
+                NonMutatingUse(NonMutatingUseContext::SharedBorrow) |
+                NonMutatingUse(NonMutatingUseContext::UniqueBorrow) |
+                NonMutatingUse(NonMutatingUseContext::AddressOf) |
+                MutatingUse(MutatingUseContext::Borrow) |
+                MutatingUse(MutatingUseContext::AddressOf) => Some(copy_fn),
+
+                NonMutatingUse(NonMutatingUseContext::Projection) |
+                MutatingUse(MutatingUseContext::SetDiscriminant) |
+                MutatingUse(MutatingUseContext::Deinit) |
+                MutatingUse(MutatingUseContext::Yield) |
+                MutatingUse(MutatingUseContext::Drop) |
+                MutatingUse(MutatingUseContext::Projection) |
+                MutatingUse(MutatingUseContext::Retag) |
+                NonUse(_) => None
+            } {
+                self.loc(location, location, ptr_fn)
+                    .arg_addr_of(deref2_base)
+                    .add_to(self);
+            }
+
+            if have_deref2 {
+                // Add an event for the pointer of the second dereference
+                self.loc(location, location, load_value_fn)
+                    .arg_var(deref2_base.clone())
+                    .add_to(self);
```

Comment:
Might be good to have an example here as well.  I wrote a big comment explaining why emitting `LoadValue` here is wrong, before finally realizing that it's correct :)